### PR TITLE
Add support for Enum by name

### PR DESCRIPTION
**Pull request by Julian-Brendel**
Following the pattern oulined by Samuel in https://github.com/samuelcolvin/pydantic/issues/598#issuecomment-503032706, one can configure Pydantic to allow Enum lookup by enum field name.

This however is not compatible with `pydantic-cli`, as it always populates argparses choices with the enum values, without any room for configuration. Argparse then fails the parsing before the value hit's the pydantic validation.

Problematic piece:
```python
choices: T.Optional[T.List[T.Any]] = None
try:
    if issubclass(field.type_, Enum):
        choices = [x.value for x in field.type_.__members__.values()]
except TypeError:
    pass
```

Proposal to use field extra kwarg `use_enum_names`, switching the choices from values to names:
```python
class Options(BaseModel):
    favorite_animal: Animal = Field(..., use_enum_names=True)
    animals: Set[Animal] = Field(..., use_enum_names=True)
```

```python
choices: T.Optional[T.List[T.Any]] = None
try:
    if extra.get("use_enum_names") is True:
        choices = [x for x in field.type_.__members__.keys()]
    elif issubclass(field.type_, Enum):
        choices = [x.value for x in field.type_.__members__.values()]
except TypeError:
    pass
```

Additionally, replaces deprecated `tests_require` (since setuptools 41.5.0) with another `extras_requires`

### CODE REVIEW

**Comment by mpkocher** [comment]
Thanks for the PR and interest in pydantic-cli. 

This PR looks like a reasonable solution, however, I think it might be useful to try to get to root issue. 

I believe you've pointed out a fundamental design issue with the original implementation. Specifically, the core issue is the duplication of the argparse layer validation and the pydantic layer validation. This friction point is creating inconsistencies and unexpected behavior. 

Pydantic has a pretty heavy-handed and opinionated approach to casting. By design, pydantic also allows a user to hook into the validation/casting (as you've demonstrated in your example). This customization won't necessarily be consistent a simple `choices` model at the argparse level. 

I would propose:

- Bypass the argparse level validation completely by removing the 'choice=' in the `add_argument` usage (This yields pydantic's casting/validation layer as the single abstraction responsible for all casting/validation)
-  Add a terse description of the Enum (using the `.value`) to the core description string.

Also, if a user wants to better communicate other allowed "castable" values at the CLI help level, this can be performed by customizing the `Field(..., description="")`. This also removes the need for a `use_enum_names` configuration parameter.

Does this seem reasonable?

**Comment by mpkocher** [comment]
I experimented with this idea a bit using one of the examples.

https://github.com/mpkocher/pydantic-cli/blob/master/pydantic_cli/examples/simple_with_enum.py

```python
from enum import Enum, auto, IntEnum
from typing import Set
from pydantic import BaseModel, Field

from pydantic_cli import run_and_exit

class CastAbleEnum:
    """Example enum mixin that will cast from case-insensitive"""

    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        try:
            lookup = {k.lower(): item.value for k, item in cls.__members__.items()}
            return lookup[v.lower()]
        except KeyError:
            raise ValueError(f"Invalid value {v}. {cls.cli_help()}")

    @classmethod
    def cli_help(cls) -> str:
        return f"Allowed={list(item.name for k, item in cls.__members__.items())}"


class Mode(CastAbleEnum, IntEnum):
     alpha = auto()
    beta = auto()

    @classmethod
    def cli_help(cls) -> str:
        return f"Allowed={list(item.name for k, item in cls.__members__.items())}"


class State(CastAbleEnum, str, Enum):
     RUNNING = "RUNNING"
    FAILED = "FAILED"
    SUCCESSFUL = "SUCCESSFUL"


class Options(BaseModel):
    states: Set[State] = Field(..., description=f"States to filter on. {State.cli_help()}")
    mode: Mode = Field(..., description=f"Processing Mode to select. {Mode.cli_help()}")
    max_records: int = 100
```

Usage:

```bash
python pydantic_cli/examples/simple_with_enum.py --help
usage: simple_with_enum.py --states STATES [STATES ...] --mode MODE [--max_records MAX_RECORDS] [--help] [--version]

optional arguments:
  --states STATES [STATES ...]
                        States to filter on. Allowed=['RUNNING', 'FAILED', 'SUCCESSFUL'] (type:State required=True)
  --mode MODE           Processing Mode to select. Allowed=['alpha', 'beta'] (type:Mode required=True)
  --max_records MAX_RECORDS
                        (type:int default:100)
  --help                Print Help and Exit
  --version             show program's version number and exit
```

Case Insensitive 

```bash
> python pydantic_cli/examples/simple_with_enum.py --states failed --mode alpha                                                                                                                                                  
Mock example running with states={'FAILED'} mode=1 max_records=100
python pydantic_cli/examples/simple_with_enum.py --states FAILED running --mode beta
Mock example running with states={'RUNNING', 'FAILED'} mode=2 max_records=100
```

And with an error:

```bash
python pydantic_cli/examples/simple_with_enum.py --states failed --mode DRAGON
1 validation error for Options
mode
  Invalid value DRAGON. Allowed=['alpha', 'beta'] (type=value_error)  File "/Users/mkocher/miniconda3/envs/core/lib/python3.9/site-packages/pydantic_cli/__init__.py", line 509, in _runner
    opts = cls(**pure_d)
  File "pydantic/main.py", line 331, in pydantic.main.BaseModel.__init__
```


There's a bit of boilerplate with a 'CastAbleEnum' to enable a custom casting/validation (e.g., using name, case insensitive) and there's some required effort to wire the custom help into the `description` of `Field`.

**Comment by Julian-Brendel** [comment]
Im a big fan of this, indeed felt odd to have both argparse and pydantic validation.
Implementation is looking good, let me change up the PR and tests.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # CHANGELOG
 
+## Version 4.3.0
+
+- Leverage Pydantic validation for enum choices, enabling more complex use-cases
+
```
Comment (mpkocher) [inline CHANGELOG.md:6]: Can you also bump the version in `_version.py` to `4.3.0`?

https://github.com/mpkocher/pydantic-cli/blob/master/pydantic_cli/_version.py

File: pydantic_cli/tests/test_examples_simple_with_enum_by_name.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+class TestExamples(_TestHarness[Options]):
+
+    CONFIG = HarnessConfig(Options, example_runner)
+
+    def test_simple_01(self):
+        args = ["--states", "RUNNING", "FAILED", "--mode", "alpha"]
+        self.run_config(args)
+
+    def test_case_insensitive(self):
+        args = ["--states", "successful", "failed", "--mode", "ALPHA"]
```
[the remark this answers is not in the document]
Comment (Julian-Brendel) [reply pydantic_cli/tests/test_examples_simple_with_enum_by_name.py:15]: Done

File: CHANGELOG.md (hunk 1 above)
Comment (Julian-Brendel) [reply CHANGELOG.md:6]: Done

**Comment by mpkocher** [comment]
Thanks for the PR and your efforts.

**Pull request merged**